third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    /// Transforms a bytecode
+    fn transform_bytecode(&mut self, bytecode: Bytecode) {
+        match bytecode {
+            Bytecode::Branch(attr_id, l0, l1, t) => self.transform_branch(attr_id, l0, l1, t),
+            // Edge of a `Jump` can never be critical since the source node only have one out edge
+            _ => self.emit(bytecode),
+        }
+    }
+
+    /// Transforms a branch instruction by splitting the critical out edges
+    pub fn transform_branch(&mut self, attr_id: AttrId, l0: Label, l1: Label, t: TempIndex) {
+        match (
+            self.split_critical_edge(attr_id, l0),
+            self.split_critical_edge(attr_id, l1),
+        ) {
+            (None, None) => self.emit(Bytecode::Branch(attr_id, l0, l1, t)),
+            (None, Some((l1_new, code))) => {
+                self.emit(Bytecode::Branch(attr_id, l0, l1_new, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code)), None) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code0)), Some((l1_new, code1))) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1_new, t));
+                self.emit_codes(code0);
+                self.emit_codes(code1);
+            },
+        }
+    }
+
+    /// `label` is the target of some branch instruction
+    /// If label has multiple sources, returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    /// otherwise returns `None`
+    fn split_critical_edge(
+        &mut self,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        // label is in `srcs_count` by construction
+        if *self.srcs_count.get(&label).expect("srcs count") > 1 {
+            Some(self.split_edge(attr_id, label))
+        } else {
+            None
+        }
+    }
+
+    /// Returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    fn split_edge(&mut self, attr_id: AttrId, label: Label) -> (Label, Vec<Bytecode>) {
+        let new_label = self.gen_fresh_label();
+        let code = vec![
+            Bytecode::Label(attr_id, new_label),
+            Bytecode::Jump(attr_id, label),
+        ];
+        (new_label, code)
+    }
+
+    /// Generates a fresh label
+    fn gen_fresh_label(&mut self) -> Label {
+        let new_label = Label::new(
+            if self.labels.is_empty() {
+                0
+            } else {
+                self.labels.iter().next_back().expect("label").as_usize() + 1
+            },
+        );
+        self.labels.insert(new_label);
+        new_label
+    }
+
+    fn emit(&mut self, bytecode: Bytecode) {
+        self.data.code.push(bytecode)
+    }
+
+    fn emit_codes(&mut self, code: Vec<Bytecode>) {
```

Comment:
nit: more general (and preferred) to have `&[Bytecode]`?

We could also get rid of the "special case" of having only code to emit, above.

Reply:
@vineethk I disagree.  With `Vec<Bytecode>` here the iterator below can reuse the instr.  With a ref param we have to copy/clone the instr.

Reply:
Fair point. I was applying a general Rust guideline of preferring slices, but I see now that it is not applicable here.